analysis/src/SignatureHelp.ml

```diff
@@ -0,0 +1,213 @@
+  let posBeforeCursor = (fst pos, max 0 (snd pos - 1)) in
+  let foundFunctionApplicationExpr = ref None in
+  let setFound r = foundFunctionApplicationExpr := Some r in
+  let expr (iterator : Ast_iterator.iterator) (expr : Parsetree.expression) =
+    (match expr with
+    | {
+     pexp_desc = Pexp_apply (({pexp_desc = Pexp_ident _} as exp), args);
+     pexp_loc;
+    }
+      when pexp_loc |> Loc.hasPosInclusive ~pos:posBeforeCursor ->
+      (* TODO: Move extractExpApplyArgs to a shared module that doesn't look like it's only for completion. *)
+      let extractedArgs = CompletionFrontEnd.extractExpApplyArgs ~args in
+      let argAtCursor =
+        let unlabelledArgCount = ref 0 in
+        extractedArgs
+        |> List.find_map (fun arg ->
+               match arg.CompletionFrontEnd.label with
+               | None ->
+                 let currentUnlabelledArgCount = !unlabelledArgCount in
+                 unlabelledArgCount := currentUnlabelledArgCount + 1;
+                 (* An argument without a label is just the expression, so we can use that. *)
+                 if arg.exp.pexp_loc |> Loc.hasPosInclusive ~pos:posBeforeCursor
+                 then Some (Unlabelled currentUnlabelledArgCount)
+                 else None
+               | Some {name; posStart; posEnd} ->
+                 (* Check for the label identifier itself having the cursor *)
+                 let cursorIsWithinLabelIdPos =
+                   posStart <= pos && pos <= posEnd
+                 in
+                 (* Check for the expr assigned to the label having the cursor *)
+                 let cursorIsWithinArgExprPos =
+                   arg.exp.pexp_loc |> Loc.hasPosInclusive ~pos:posBeforeCursor
+                 in
+                 (* This most likely means that parsing failed and recovered. Happens for empty assignments for example. *)
+                 let argExprIsParserRecovery =
+                   arg.exp.pexp_loc |> Loc.end_ = (Location.none |> Loc.end_)
+                 in
+                 if
+                   cursorIsWithinLabelIdPos || cursorIsWithinArgExprPos
+                   || argExprIsParserRecovery
+                 then Some (Labelled name)
+                 else None)
+      in
+      setFound (argAtCursor, exp, extractedArgs)
+    | _ -> ());
+    Ast_iterator.default_iterator.expr iterator expr
+  in
+  let iterator = {Ast_iterator.default_iterator with expr} in
+  let parser = Res_driver.parsingEngine.parseImplementation ~forPrinter:false in
+  let {Res_driver.parsetree = structure} = parser ~filename:currentFile in
+  iterator.structure iterator structure |> ignore;
+  match !foundFunctionApplicationExpr with
+  | Some (argAtCursor, exp, extractedArgs) -> (
+    (* Not looking for the cursor position after this, but rather the target function expression's loc. *)
+    let pos = exp.pexp_loc |> Loc.end_ in
+    let completions, env, package =
+      let textOpt = Files.readFile currentFile in
```

Comment:
This is where one switches to use the parser right?

Reply:
Correct.